chore(backend): add TLS certificate rotation documentation and helper scripts. Fixes #12328

rahul810050 opened the pull request:
chore(backend): add TLS certificate rotation documentation and helper scripts. Fixes #12328

**Description of your changes:**

This PR adds missing documentation for TLS certificate rotation required when using the pod-to-pod TLS feature introduced in PR #12082. When TLS secrets are renewed, backend services (API server, persistence agent, metadata writer, cache server, etc.) do not automatically reload updated certificate data. A rolling restart is required. This behavior was previously undocumented.

The updates included in this PR:

1. **Documentation update (`backend/README.md`):**
   - Added a new section **“TLS Certificate Rotation (Pod-to-Pod TLS)”**.
   - Explained why certificate rotation is needed, how secrets interact with backend components, and why pods must be restarted.
   - Added a complete, copy-paste-ready rotation procedure:
     - Generate/obtain new TLS certs (`server.crt` / `server.key`)
     - Update the TLS secret using `kubectl create secret tls ... | kubectl apply -f -`
     - Restart affected deployments with `kubectl rollout restart`
     - Verify rollouts and confirm certificate is active.
   - Added best practices, troubleshooting notes, common errors, and automation guidance (cert-manager + checksum annotations).
   - Added cluster discovery commands to help users identify the exact secret and deployments referencing it.

2. **Helper scripts added (optional but helpful for operators):**
   - `scripts/find-tls-refs.sh`  
     Identifies which pods/deployments reference the TLS secret (via volumes or env secretKeyRef).  
     Helps operators know exactly which deployments must be restarted.
   - `scripts/rotate-tls.sh`  
     Applies new TLS cert/key to the Kubernetes Secret and automatically restarts all deployments referencing the secret, waiting for rollout completion.

3. **General improvements:**
   - Normalized README formatting.
   - Ensured all examples are reproducible and use correct fenced code blocks.
   - Provided safer and clearer operational guidance for cluster administrators.

These changes directly address the missing operational documentation noted in issue #12328 and align with the maintainers’ suggestion to include example commands.

---

**Checklist:**
- [ ] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [ ] The title for your pull request (PR) follows the title convention.  
      [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
    * `chore: set up changelog generation tools`
    * `test: fix CI failure. Part of #1234`
-->

---

**Links**
- Related feature PR (TLS implementation): #12082  
- Documentation issue: #12328

rahul810050 committed:
docs(backend): add TLS certificate rotation documentation and helper scripts. Fixes #12328

Signed-off-by: rahul810050 <rahul810050@gmail.com>

alyssacgoins commented on backend/README.md:
```diff
@@ -1,359 +1,194 @@
+### Context
 
-You can also access the Kubeflow Pipelines web interface at [http://localhost:3000](http://localhost:3000).
+When pod-to-pod TLS is enabled (see PR #12082), backend components (API server, launcher, persistence agent, cache, metadata writer, etc.) use TLS certificates stored in Kubernetes Secret(s). These certificates expire and must be rotated periodically. Updating the Secret object does not automatically make running pods load the new certificate: a restart (rolling restart) of the affected deploymen ...
```
You don't need to reference the PR here

alyssacgoins commented on backend/README.md:
````diff
@@ -1,359 +1,194 @@
-```bash
-mysql -h 127.0.0.1 -u root
-```
+### Which secrets and components are typically impacted
````
You can remove "typically", as the same deployments will be impacted each time

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,124 @@ Run the following to delete the cluster (once you are finished):
+
+### Context
+
+When pod-to-pod TLS is enabled, backend components (API server, launcher, persistence agent, metadata writer, and other pipeline-related deployments) use TLS certificates stored in Kubernetes Secret(s). These certificates expire and must be rotated periodically.
```
double-check the components listed in this line, should be a previous comment on this as well.

alyssacgoins commented on scripts/find-tls-refs.sh:
```diff
@@ -0,0 +1,30 @@
+# Defaults: namespace=kubeflow, secret=kfp-pod-tls
+
+NS="${1:-kubeflow}"
+SECRET="${2:-kfp-pod-tls}"
```
default secret name is `kfp-api-tls-cert`, as referenced here - https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/env/cert-manager/base-tls-certs/kfp-api-cert.yaml#L6

rahul810050 committed:
Update README.md

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 committed:
Update find-tls-refs.sh

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 committed:
Update rotate-tls.sh

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 committed:
Update README.md

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 commented on scripts/find-tls-refs.sh:
```diff
@@ -0,0 +1,30 @@
+# Defaults: namespace=kubeflow, secret=kfp-pod-tls
+
+NS="${1:-kubeflow}"
+SECRET="${2:-kfp-pod-tls}"
```
Yes — while `kfp-api-tls-cert` is the default in the official TLS-enabled manifests, 
there are valid scenarios where operators override the secret name:

• clusters using custom cert-manager issuers  
• users providing externally-issued enterprise certificates  
• users overriding the secret via kustomize patches  
• dev environments that use a different name (e.g., kfp-pod-tls)

Because of these scenarios, I kept the argument optional — but updated the default 
to `kfp-api-tls-cert` as you suggested.

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,121 @@ Run the following to delete the cluster (once you are finished):
+
+### Context
+
+When pod-to-pod TLS is enabled, backend components (ml-pipeline-apiserver, metadata-envoy-deployment, metadata-grpc-deployment, ml-pipeline-persistenceagent, ml-pipeline-scheduledworkflow, and ml-pipeline-ui) use the TLS certificate stored in the Kubernetes TLS Secret `kfp-api-tls-cert`. These certificates expire and must be rotated periodically.
```
```suggestion
When pod-to-pod TLS is enabled, backend components (`ml-pipeline-apiserver`, `metadata-envoy-deployment`, `metadata-grpc-deployment`, `ml-pipeline-persistenceagent`, `ml-pipeline-scheduledworkflow`, and `ml-pipeline-ui`)
For readability
```

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,121 @@ Run the following to delete the cluster (once you are finished):
+* `ml-pipeline-ui`
+
+> **Note**
+> The `pipelines-cache` / cache deployment generally does **not** require the CA cert for inbound TLS from pipeline components; confirm whether your deployment references the TLS secret before restarting it.
```
Rename/syntax:
```suggestion
> The `cache-server`  deployment generally does **not** require the CA cert for inbound TLS from pipeline components; confirm whether your deployment references the TLS secret before restarting it.
```

alyssacgoins reviewed (changes requested):
Nice work! Just a couple small changes. I tested your scripts locally and everything runs smoothly.

rahul810050 committed:
Clarify TLS certificate rotation details in README

Updated wording for clarity regarding TLS certificate rotation and cache server deployment.

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 committed:
Enhance TLS and Secret update instructions in README

Updated the README to include instructions for confirming the TLS certificate and automating pod restarts after a Secret update.

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 commented:
@alyssacgoins I have updated the pr according to your feedback. Please feel free to review it whenever you get a chance

Thanks!!

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,154 @@ Run the following to delete the cluster (once you are finished):
+
+### Recommended rotation procedure
+
+#### Prepare/obtain new cert and key
```
```suggestion
#### 1. Prepare/obtain new cert and key
```
Numbering the steps would add some readability to these instructions

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,154 @@ Run the following to delete the cluster (once you are finished):
+#### Prepare/obtain new cert and key
+Generate or obtain new cert files: `server.crt` and `server.key` (PEM encoded).
+
+#### Update the Kubernetes TLS secret
```
```suggestion
#### 2. Update the Kubernetes TLS secret
```

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,154 @@ Run the following to delete the cluster (once you are finished):
+
+Or restart all pipeline-related deployments discovered earlier.
+
+#### Verify rollout and pod readiness
```
```suggestion
#### 4. Verify rollout and pod readiness
```

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,154 @@ Run the following to delete the cluster (once you are finished):
+* **Auditing**: record rotation times and certificate fingerprints for traceability.
+* **Validate**: after restarts, verify connectivity between components and watch logs for TLS handshake issues.
+
+## Optional: automate restart after secret update
```
If Kubernetes as a rule does not automatically restart pods, why is this step optional? I think it can potentially confuse users. You could add this section as a 5th step under **Recommended rotation procedure**

Since Kubernetes as a general rule does not automatically restart pods after a secret is updated, it will likely be confusing to users if this step is labeled **optional**. Are there any cases where a user would not need to restart the pods after?
It might make more sense add this section as a 5th step under **Recommended rotation procedure**.

alyssacgoins commented on backend/README.md:
```diff
@@ -349,6 +349,154 @@ Run the following to delete the cluster (once you are finished):
+
+## Best practices & notes
+
+* Rotation interval: rotate certificates before they expire; recommended: track expiry via:
```
```suggestion
* Rotation interval: rotate certificates before they expire. Track expiry with the following command:
```

alyssacgoins commented on backend/README.md:
````diff
@@ -349,6 +349,154 @@ Run the following to delete the cluster (once you are finished):
+  | grep -C3 "name: <secret-name>" || true
+```
+
+### Recommended rotation procedure
````
```suggestion
### Certificate rotation procedure
```

rahul810050 committed:
Revise TLS certificate renewal process in README

Updated TLS certificate renewal documentation for clarity and structure.

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 committed:
Update backend/README.md

Co-authored-by: Alyssa Goins <80764587+alyssacgoins@users.noreply.github.com>
Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 committed:
Document TLS certificate rotation helper scripts

Added helper scripts for TLS certificate rotation procedures.

Signed-off-by: Rahul Naskar <143400898+rahul810050@users.noreply.github.com>

rahul810050 commented:
@alyssacgoins could you please review it??
